Refactor the way Split Update nodes are constructed in the planner.

hlinnaka opened the pull request:
First commit:
```
Refactor the way Split Update nodes are constructed in the planner.

One specialty in a Split Update is that the node needs the *old* values
for all the updated rows, to compute the distribution hash for each
old row, so that they can be deleted. That was previously handled at
the time when the SplitUpdate node was created, by adding any missing
Vars for the old values to the subplan's target list, pushing them down
through joins and any other plan nodes, all the way down to the Scan
node for that relation. That seemed complicated and fragile.

The reason to tackle this right now is that we were seeing failures related
to this, while working on the PostgreSQL 9.4 merge. It added a test case,
where a Split Update was done through a security barrier view. The security
barrier view added a SubqueryScan to the plan tree, and the mechanism to
push through the old attributes couldn't cope with that. I'm sure we
could've hacked that to make it work, but this refactoring seems like a
better long term fix.

This patch makes it the responsibility of preprocess_targetlist(), to ensure
that the old values are made available to the top of the tree, if a Split
Update is needed. preprocess_targetlist() seems like the appropriate place,
because it already does that for columns that are not modified by the
UPDATE.

Now that we are making the decision on whether to do a split update in
preprocess_targetlist() already, add a flag to PlannerInfo to remember that
decision, to the point where the ModifyTable node is added to the top of
the plan tree.

Also add a test case, for an inherited table where some children have a
different distribution key, and an UPDATE on some of the children require a
Split Update, and others don't. That was causing me trouble at one point
during the development, and I'm not sure if there was any existing test to
cover that.
```

Second commit:
```
Move the code that adds 'gp_segment_id' column to UPDATE tlist.

Per the comment, it should be added in rewriteTargetListUD(), just like
'ctid' is.
```

kainwen commented:
Hi, @hlinnaka 

Since you are refactoring this part of the code, would you please help fix this issue in your PR?

https://github.com/greenplum-db/gpdb/issues/5882

joaopapereira commented on src/backend/cdb/cdbmutate.c:
```diff
@@ -1440,7 +1242,30 @@ find_ctid_attribute_check(List *targetList, AttrNumber *ctidAttr, Index resultRe
+	var = (Var *) (tle->expr);
+
+	/* OID should follow after normal attributes */
+	Assert((var->varno == resultRelationsIdx &&
```
(This is more for my education than anything else) Does this assert do write any logs? or we do not care?

hlinnaka replied:
If an Assert fails, you get a PANIC, which shows the source line where Assert was. The PANIC willl reset the whole server. Yeah, it's also logged to the server log. Did that answer your question?

While we're at it: IMHO we should almost always use `elog(ERROR, ...)`, instead of Assert in the planner. If you hit this Asset, for example, it's almost certainly because the planner messed up something, forgot to pass through a column through some node or something like that. If you happen to hit that in production, you probably don't want to just ignore it, which is what happens in production builds at the moment. Asserts are compiled away in production builds, but in the planner, this kinds of checks are practically free, there's no need to try to optimize them away. OTOH, you surely don't want to PANIC and bring down the server, either. If you just ERROR out, and abort the transaction, the system should be fine to run other queries. An `elog(ERROR, ...` seems just right

I didn't do that here, since I just copied the code from the find_ctid_attribute_check(). Come to think of it, I probably should have..

hlinnaka commented:
I changed this some more, based on some problems we were seeing on the 9.4 merge iteration branch with this. Before, the SplitUpdate executor node required that *all* the old values must be present in the target list. But that's more than we need; we actually only need the old values for distribution key columns. I modified the executor node to replace NULLs for any columns for which the old value is not available. That allowed some further simplifications: Instead of keeping track of which columns were changed and which not, and only adding the old value of changed columns to the target list, we now add the old values of every distribution key column to the target list, right after all the new values. This in turn eliminates the code that was previously needed to find the old values in the target list: we now know that they come immediately after the new values.

Please take another look.

hlinnaka commented:
@kainwen: This is quite unrelated to issue #5882. I agree that one is a genuine problem, though. We should address that, but I'm afraid it's going to require much bigger changes.

asimrp commented on src/test/regress/expected/update_gp.out:
```diff
@@ -70,6 +70,35 @@ update target set b = 10 where c = 10;
+         Hash Key: ((child_a.a + 1))
+         ->  Split
+               ->  Seq Scan on child_a
+   ->  Explicit Redistribute Motion 3:3  (slice3; segments: 3)
```
This is surprising to me.  Why redistribute `child_a` when it's distribution key is not getting updated?

hlinnaka replied:
The distribution key of `child_a` *is* being updated here. I think you meant `child_b`.

Hmm. It's true that the Explicit Redistribute Motion is superfluous. In a more complicated case, though, the source rows might come from different segments, and the Motion is required. In simple cases, where the plan contains no other Motion nodes, the planner elides the creation of the Explicit Motion at a later stage (https://github.com/greenplum-db/gpdb/blob/master/src/backend/cdb/cdbmutate.c#L792), but that optimization doesn't kick in in this statement, because of the Redistribute Motions in the other subplans. That's a bit unfortunate, perhaps we could work a bit harder to determine if the current flow matches the distribution of the target table. But this isn't a new issue in this PR, so I'm not to try fixing it now.

hlinnaka commented:
Pushed. No one explicitly +1'd this, but several people at least looked at the patch, and didn't raise concerns. This has been sitting for a long time, so I'll take that as approval.
